contracts/Whitelist.sol

```diff
@@ -9,24 +10,54 @@ import "./packages/oz/Ownable.sol";
  * @notice The whitelist module keeps track of all valid Otoken contracts.
  */
 contract Whitelist is Ownable {
-    ///@notice mapping to track whitelisted product
+    ///@notice AddressBook module address
+    address public addressBook;
+
+    ///@dev mapping to track whitelisted products
     mapping(bytes32 => bool) internal whitelistedProduct;
+    ///@dev mapping to track whitelisted otokens
+    mapping(address => bool) internal whitelistedOtoken;
+
+    /**
+     * @dev constructor
+     * @param _addressBook AddressBook module address
+     */
+    constructor(address _addressBook) public {
+        require(_addressBook != address(0), "Invalid address book");
 
-    ///@notice emitted when owner whitelist a product
+        addressBook = _addressBook;
+    }
+
+    ///@notice emitted when the owner whitelists a new product
     event ProductWhitelisted(
         bytes32 productHash,
         address indexed underlying,
         address indexed strike,
         address indexed collateral
     );
 
+    ///@notice emitted when Otoken Factory module whitelist an otoken
+    event OtokenWhitelisted(address otoken);
+
     /**
-     * @notice check if a product is supported
+     * @notice check if sender is Otoken Factory module
+     */
+    modifier onlyFactory() {
+        require(
+            msg.sender == AddressBookInterface(addressBook).getOtokenFactory(),
+            "WhiteList: Sender is not Otoken Factory"
+        );
+
+        _;
+    }
+
+    /**
+     * @notice check if a product is whitelisted
      * @dev product = the hash of underlying, strike and collateral asset
      * @param _underlying option underlying asset address
      * @param _strike option strike asset address
      * @param _collateral option collateral asset address
```

Comment:
update these with the comment @aleone had on https://github.com/opynfinance/GammaProtocol/pull/6